r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1140 @@
+```{r, include = FALSE}
+neighborhoods_2 <- neighborhoods %>% 
+  rename(nta_code = nta)
+```{r}
+neighborhoods_2 
+```
+
+Then attempting the join will result in an error - `data sources have no common variables`:
+```{r, error = TRUE}
+dogs %>% left_join(neighborhoods_2)
+```
+You can use the `by` argument to the join function to specify which columns are to be used to do the join:
+```{r}
+dogs %>% 
+  left_join(neighborhoods_2, by = c("nta" = "nta_code"))
+```
+The value passed to `by` is a named character vector and is created using the form
+```
+c("<< Name of column in LHS dataset >>" = "<< Name of column in RHS dataset >>")
+```
+The `by` argument in the example above, 
+```{r, results = "hide"}
+c("nta" = "nta_code")
+```
+specifies the join column is `"nta"` in `dogs`, and `"nta_code"` in `neighborhoods_2`.
+
+### Exercise: Predicting the output of a join
+
+Try to predict what this code will return, then run it and check your prediction:
+```{r, results = "hide", message = FALSE}
+neighborhoods %>% left_join(dogs)
+```
+
+### Exercise: Dog licenses per capita
+
+* Use `left_join()` to join `dogs_by_nta` with `nta_pop`. 
+
+* `filter()` the result to only use population from `year` 2010.
+
+* Add a `mutate()` step to find the number of dog licenses per person in each NTA.
+
+* Which neighborhood has the highest dog licenses per person based on the 2010 population?
+
+```{r, include = FALSE}
+dogs_by_nta %>% 
+  left_join(nta_pop, by = c("neighborhood_tabulation_area" = "nta_code")) %>% 
+  mutate(licenses_per_person = n_licences/population) %>% 
+  filter(year == 2010) %>% 
+  arrange(desc(licenses_per_person)) %>% 
+  select(nta_name, licenses_per_person)
+# Stuyvesant Town-Cooper Village averages 0.045 dogs per person
+# Or ~45 dogs per 1000 people
+```
+
+### Other join functions
+
+Like `left_join()` all the join functions in dplyr also end in `_join` and take two datasets as arguments.  You can see the full list in:
+```{r, eval = FALSE}
+?dplyr::join
+```
+
+The functions `inner_join()`, `left_join()`, `right_join()` and `full_join()` all take two datasets and return a single dataset with **all the columns** from both.  They differ in the treatment of rows with no matches:
+
+* `left_join()`, as you've seen, keeps rows in the left hand dataset, but if there is no match in the right hand dataset, columns from the right hand dataset are given missing values.
+    ```{r}
+    dogs %>% left_join(neighborhoods)
+    ```
+
+* `right_join()`, is just like `left_join()` but with the roles of the data sets reversed.
+    ```{r}
+    dogs %>% right_join(neighborhoods)
+    ```
+
+* `inner_join()` returns only rows that have matches.
```

Comment:
We're going to have to test-drive this with a few learners - again, it feels like a leap in complexity.